crates/shared/src/token_info.rs

```diff
@@ -1,173 +1,239 @@
 use {
-    crate::ethrpc::Web3,
+    crate::{ethcontract_error::EthcontractErrorType, ethrpc::Web3},
+    anyhow::Result,
     async_trait::async_trait,
     contracts::ERC20,
-    ethcontract::{batch::CallBatch, H160},
-    std::{collections::HashMap, sync::Arc},
-    tokio::sync::Mutex,
+    ethcontract::{errors::MethodError, H160},
+    futures::{
+        future::{BoxFuture, Shared},
+        FutureExt,
+    },
+    model::order::BUY_ETH_ADDRESS,
+    std::{
+        collections::HashMap,
+        sync::{Arc, Mutex},
+    },
+    thiserror::Error,
 };
 
-const MAX_BATCH_SIZE: usize = 100;
-
 #[cfg_attr(test, derive(Eq, PartialEq))]
 #[derive(Clone, Debug, Default)]
 pub struct TokenInfo {
     pub decimals: Option<u8>,
     pub symbol: Option<String>,
 }
 
-pub struct TokenInfoFetcher {
-    pub web3: Web3,
-}
+#[derive(Clone, Debug, Error)]
+#[error("error fetching token info: {0}")]
+pub struct Error(String);
 
 #[mockall::automock]
 #[async_trait]
 pub trait TokenInfoFetching: Send + Sync {
+    /// Retrieves information for a token.
+    async fn get_token_info(&self, address: H160) -> Result<TokenInfo, Error>;
+
     /// Retrieves all token information.
     /// Default implementation calls get_token_info for each token and ignores
     /// errors.
     async fn get_token_infos(&self, addresses: &[H160]) -> HashMap<H160, TokenInfo>;
 }
 
+pub struct TokenInfoFetcher {
+    pub web3: Web3,
+}
+
+impl TokenInfoFetcher {
+    async fn fetch_token(&self, address: H160) -> Result<TokenInfo, Error> {
+        if address == BUY_ETH_ADDRESS {
+            let chain = self
+                .web3
+                .eth()
+                .chain_id()
+                .await
+                .map_err(|err| Error(err.to_string()))?;
+            return Ok(TokenInfo {
+                decimals: Some(18),
+                symbol: match chain.as_u64() {
+                    1 | 5 => Some("ETH".to_string()),
+                    100 => Some("xDAI".to_string()),
+                    _ => None,
```

Comment:
We should probably error more loudly here for when we go on other chains.

Reply:
How important is the symbol here? Is it sufficient to just call it _NativeAsset_ or the like?

Reply:
It isn't important at all. I like your suggestion a lot and will change it.

Reply:
6c3aedd5c95a456c4c27f6b815ecaffb3f99f4a5